Fix issue #9858.

jiefangxuanyan opened the pull request:
Added some byte swapping code to solve endianness problem.

jiefangxuanyan committed:
Fix issue #9858.

tensorflow-jenkins commented:
Can one of the admins verify this patch?

jiefangxuanyan commented:
I signed it!

rmlarsen commented on tensorflow/core/kernels/decode_raw_op.cc:
```diff
@@ -73,6 +73,13 @@ class DecodeRawOp : public OpKernel {
     for (int64 i = 0; i < flat_in.size(); ++i) {
       const T* in_data = reinterpret_cast<const T*>(flat_in(i).data());
       memcpy(out_data, in_data, str_size);
+      if (port::kLittleEndian != little_endian_) {
```
Instead if memcpy + byteswap, just use a custom copy loop for the big endian case, so you don't have to touch the data twice.

jiefangxuanyan committed:
Write a custom copy loop to avoid touching the data twice.

jiefangxuanyan commented:
@rmlarsen like this?

rmlarsen commented on tensorflow/core/kernels/decode_raw_op.cc:
```diff
@@ -70,10 +70,24 @@ class DecodeRawOp : public OpKernel {
+        memcpy(out_data, in_data, str_size);
+        out_data += added_dim;
+      }
+    } else {
```
Please add unit test(s) such that both branches are tested.

rmlarsen reviewed (changes requested):
Please add unit test(s).

rmlarsen reviewed (changes requested):
Please add a unit test.

jiefangxuanyan committed:
Add a criterion for 1-byte data types.

jiefangxuanyan committed:
Add test for both endianness.

jiefangxuanyan commented:
I added a test case for endianness. By the way, according to the documentation (https://www.tensorflow.org/versions/master/api_docs/python/tf/decode_raw), I think this op should behave consistently on platforms with different endianness, so I removed the endianness test in an existing test case. If the original behavior in that test case is expected, maybe you should make changes on the documentation.

rmlarsen commented on tensorflow/python/kernel_tests/decode_raw_op_test.py:
```diff
@@ -54,18 +54,26 @@ def testToInt16(self):
+          in_bytes, out_type=dtypes.int32, little_endian=True)
+      decode_be = parsing_ops.decode_raw(
+          in_bytes, out_type=dtypes.int32, little_endian=False)
+      result = decode_le.eval(feed_dict={in_bytes: ["\x01\x02\x03\x04"]})
```
Won't this depend on the endianness of the machine where this test is run? You can test the endianness of the machine here using the Python sys module, e.g.,
```
>>> import sys
>>> sys.byteorder
'little'
```

jiefangxuanyan replied:
It won't. This decode op gets same result on machines with different endianness, just like the way these machines receive data from the network. In fact, platform endianness is considered in the C++ code (`if (port::kLittleEndian == little_endian_)`) to ensure consistency. As I said before, the old platform-dependent behavior in this test is contradictionary with the documentation.

rmlarsen commented:
@tensorflow-jenkins test this please
